core/flyout_vertical.js

```diff
@@ -227,36 +229,73 @@ Blockly.VerticalFlyout.prototype.wheel_ = function(e) {
  * @protected
  */
 Blockly.VerticalFlyout.prototype.layout_ = function(contents, gaps) {
-  this.workspace_.scale = this.targetWorkspace.scale;
+  Blockly.Flyout.prototype.layout_.call(this, contents, gaps);
+  this.layoutContent_();
+};
+
+/**
+ * Respositions the blocks, buttons and labels in the flyout based on their
+ *     current dimensions and order.
+ * @protected
+ */
+Blockly.VerticalFlyout.prototype.layoutContent_ = function() {
+  this.workspace_.setResizesEnabled(false);
+  this.reflow();
   var margin = this.MARGIN;
-  var cursorX = this.RTL ? margin : margin + this.tabWidth_;
+  var cursorX = margin + this.tabWidth_;
   var cursorY = margin;
 
+  var contents = this.contents_.slice();
+  var gaps = this.gaps_.slice();
+  
   for (var i = 0, item; (item = contents[i]); i++) {
-    if (item.type == 'block') {
-      var block = item.block;
-      var allBlocks = block.getDescendants(false);
-      for (var j = 0, child; (child = allBlocks[j]); j++) {
-        // Mark blocks as being inside a flyout.  This is used to detect and
-        // prevent the closure of the flyout if the user right-clicks on such a
-        // block.
-        child.isInFlyout = true;
+    if (item instanceof Blockly.BlockSvg) {
+      var block = item;
+      var dimensions = block.getHeightWidth();
+      var position = block.getRelativeToSurfaceXY();
+
+      var destinationX = cursorX;
+      if (block.outputConnection && !this.RTL) {
+        destinationX -= this.tabWidth_;
       }
-      block.render();
-      var root = block.getSvgRoot();
-      var blockHW = block.getHeightWidth();
-      var moveX = block.outputConnection ? cursorX - this.tabWidth_ : cursorX;
-      block.moveBy(moveX, cursorY);
 
-      var rect = this.createRect_(block,
-          this.RTL ? moveX - blockHW.width : moveX, cursorY, blockHW, i);
+      if (this.RTL) {
+        destinationX = (this.width_ / this.workspace_.scale) - (margin * 2) +
+            (block.outputConnection ? this.tabWidth_ : 0);
+      }
+      var destinationY = cursorY;
+
+      var deltaX = destinationX - position.x;
+      var deltaY = destinationY - position.y;
+
+      if (deltaX || deltaY) {
+        block.moveBy(deltaX, deltaY);
+      }
+      this.moveRectToBlock_(block.flyoutRect_, block);
+    } else if (item instanceof Blockly.FlyoutButton) {
+      item.moveTo(cursorX, cursorY);
+      var dimensions = item;
+    }
 
-      this.addBlockListeners_(root, block, rect);
+    cursorY += dimensions.height + gaps[i];
```

Comment:
I think I would prefer this as it was before. Right now dimensions can either be a button or a `{height, width}`  which is a bit confusing and I think might cause problems later on when we try to have strict types for everything.